src/libmeasurement_kit/ooni/captive_portal.cpp

```diff
@@ -0,0 +1,231 @@
+    for (auto input : gen_http_inputs()) {
+        logger->info("setting up %s", input.at("name").c_str());
+        options["http/url"] = input.at("url");
+        std::string body;
+        http::Headers headers;
+        if (input.count("ua")) {
+            headers = {{"User-Agent", input.at("ua")}};
+        } else {
+            headers = constants::COMMON_CLIENT_HEADERS;
+        }
+
+        // Note: the following callback uses `[=]`, which means that the
+        // options are copied in each run. Thus, each test is getting its
+        // right URL to connect to and check.
+        continuations.push_back([=](Callback<Error> done_cb) {
+            templates::http_request(entry, options, headers, body,
+                                    http_cb(input, done_cb), reactor, logger);
+        });
+    }
+    mk::parallel(continuations, all_done_cb, 3);
+}
+
+// this is the only test that doesn't follow the pattern of those above.
+// this hostname should always resolve to this IP.
+void dns_msft_ncsi(Var<Entry> entry, Callback<Error> done_cb, Settings options,
+                   Var<Reactor> reactor, Var<Logger> logger) {
+    std::string hostname = "dns.msftncsi.com";
+    // Note: we're setting the nameserver to empty, which is going to work
+    // as long as we're using the `system` DNS resolver
+    std::string nameserver = "";
+    templates::dns_query(entry, "A", "IN", hostname, nameserver,
+                         [=](Error err, Var<dns::Message> message) {
+                             Entry result = {{"name", "Microsoft DNS"},
+                                             {"result", nullptr},
+                                             {"hostname", "dns.msftncsi.com"},
+                                             {"expected_ip", "131.107.255.255"},
+                                             {"failure", nullptr}};
+                             if (!!err) {
+                                 logger->info("dns_query err: %s",
+                                              err.as_ooni_error().c_str());
+                                 result["failure"] = err.as_ooni_error();
+                                 done_cb(err);
+                             } else {
+                                 for (const auto &a : message->answers) {
+                                     // XXX: This should actually be adapted
+                                     // to also deal with IPv6 replies
+                                     result["actual_ips"].push_back(a.ipv4);
+                                 }
+                                 if (message->answers.size() != 1) {
+                                     logger->info("maybe captive portal");
+                                     result["result"] = false;
+                                 } else if (message->answers[0].ipv4 !=
+                                            "131.107.255.255") {
+                                     logger->info("probably captive portal");
```

Comment:
I see @bassosimone usually prefixes logging calls with the scope in which it's happening. I think that is quite useful especially for debugging otherwise you are lost in the logs.